Cgen didnt declare type

### Describe the bug

<title>

### Reproduction Steps

```v
import time

pub type EventListener[T] = fn (T) !

struct Chan[T] {
	c chan T
}

struct EventWaiter[T] {
	check ?fn (T) bool
	c &Chan[T]
}

pub struct EventController[T] {
mut:
	id int
	wait_fors map[int]EventWaiter[T]
	listeners map[int]EventListener[T]
}

fn (mut ec EventController[T]) generate_id() int {
	return ec.id++
}

@[params]
pub struct EmitOptions {
pub:
	error_handler ?fn (int, IError)
}

pub fn (mut ec EventController[T]) emit(e T, options EmitOptions) {
}

@[params]
pub struct EventWaitParams[T] {
pub:
	check ?fn (T) bool
	timeout ?time.Duration
}

pub fn (mut ec EventController[T]) wait[T](params EventWaitParams[T]) ?T {
	return none
}

pub fn (mut ec EventController[T]) override[T](listener EventListener[T]) EventController[T] {
	return ec
}

pub fn (mut ec EventController[T]) listen[T](listener EventListener[T]) EventController[T] {
	return ec
}

fn main() {
	ec := EventController[int]{}
}
```

### Expected Behavior

Successful compile

### Current Behavior

```v
PS D:\Games\Proekti\V\interactions> v -cg run repro.v
repro.v:54:2: warning: unused variable: `ec`
   52 |
   53 | fn main() {
   54 |     ec := EventController[int]{}
      |     ~~
   55 | }
C:/Users/mclr/AppData/Local/Temp/v_0/repro.17533198570123215954.tmp.c:540: warning: WINVER redefined
C:/Users/mclr/AppData/Local/Temp/v_0/repro.17533198570123215954.tmp.c:942: warning: InterlockedIncrement64 redefined
In file included from C:/Users/mclr/AppData/Local/Temp/v_0/repro.17533198570123215954.tmp.c:1199:
d:/games/proekti/v/v/thirdparty/tcc/include/winapi/synchapi.h:13: warning: CreateEvent redefined
In file included from C:/Users/mclr/AppData/Local/Temp/v_0/repro.17533198570123215954.tmp.c:1199:
d:/games/proekti/v/v/thirdparty/tcc/include/winapi/synchapi.h:117: warning: OpenEvent redefined
In file included from C:/Users/mclr/AppData/Local/Temp/v_0/repro.17533198570123215954.tmp.c:1199:
d:/games/proekti/v/v/thirdparty/tcc/include/winapi/synchapi.h:167: warning: CreateMutex redefined
C:/Users/mclr/AppData/Local/Temp/v_0/repro.17533198570123215954.tmp.c:3051: error: invalid type
builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.3 08189d6

### Environment details (OS name and version, etc.)

```v
V full version: V 0.4.3 140c838.08189d6
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 12 cpus, 64bit, little endian, 

getwd: D:\Games\Proekti\V\interactions
vexe: D:\Games\Proekti\V\v\v.exe
vexe mtime: 2023-12-09 09:55:27

vroot: OK, value: D:\Games\Proekti\V\v
VMODULES: OK, value: C:\Users\mclr\.vmodules
VTMP: OK, value: C:\Users\mclr\AppData\Local\Temp\v_0

Git version: git version 2.37.0.windows.1
Git vroot status: weekly.2023.48-76-g08189d64
.git/config present: true

CC version:
thirdparty/tcc status: thirdparty-windows-amd64 e90c2620
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/struct.v
+++ b/vlib/v/checker/struct.v
@@ -900,6 +900,23 @@ or use an explicit `unsafe{ a[..] }`, if you do not want a copy of the slice.',
 				generic_names := struct_sym.info.generic_types.map(c.table.sym(it).name)
 				node.typ = c.table.unwrap_generic_type(node.typ, generic_names, concrete_types)
 			}
+		} else if struct_sym.info.generic_types.len > 0
+			&& struct_sym.info.generic_types.len == struct_sym.info.concrete_types.len
+			&& c.table.cur_concrete_types.len == 0 {
+			parent_type := struct_sym.info.parent_type
+			parent_sym := c.table.sym(parent_type)
+			for method in parent_sym.methods {
+				generic_names := struct_sym.info.generic_types.map(c.table.sym(it).name)
+				for i, param in method.params {
+					if i == 0 || !param.typ.has_flag(.generic) {
+						continue
+					}
+					param_sym := c.table.sym(param.typ)
+					if param_sym.kind in [.struct_, .interface_, .sum_type] {
+						c.table.unwrap_generic_type(param.typ, generic_names, struct_sym.info.concrete_types)
+					}
+				}
+			}
 		}
 	}
 	return node.typ

--- a/vlib/v/tests/generics_struct_method_params_is_generics_struct_and_not_called_test.v
+++ b/vlib/v/tests/generics_struct_method_params_is_generics_struct_and_not_called_test.v
@@ -1,0 +1,17 @@
+struct Param[T] {
+}
+
+struct Struct[T] {
+}
+
+pub fn (s Struct[T]) method[T](p Param[T]) {
+}
+
+fn test_main() {
+	_ = Struct[int]{}
+	assert true
+	// NOTE:
+	// Do not test call Struct.method() here,
+	// to test unwrap of generic struct parameters when the method is not called.
+	// test results only need to can compile.
+}

